azkaban-exec-server/src/main/java/azkaban/dag/DagBuilder.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2018 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.dag;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * A builder to build DAGs.
+ *
+ * <p>Use the {@link DagBuilder#createNode} method to create NodeBuilder instances. Call
+ * methods on NodeBuilder to add dependencies among them. Call the {@link DagBuilder#build()} method
+ * to build a Dag.
+ */
+public class DagBuilder {
+
+  private final String name;
+  private final DagProcessor dagProcessor;
+
+  private final List<NodeBuilder> builders = new ArrayList<>();
+  private final Set<String> nodeNamesSet = new HashSet<>();
+
+  public DagBuilder(final String name, final DagProcessor dagProcessor) {
+    requireNonNull(name, "The name of the DagBuilder can't be null");
+    this.name = name;
+    requireNonNull(name, "The dagProcessor of the DagBuilder can't be null");
+    this.dagProcessor = dagProcessor;
+  }
+
+  public NodeBuilder createNode(final String name, final NodeProcessor nodeProcessor) {
+    final NodeBuilder builder = new NodeBuilder(name, nodeProcessor, this);
+    this.builders.add(builder);
+    if (this.nodeNamesSet.contains(name)) {
```

Comment:
should we do existence check before builders.add(builder)?

Reply:
I don't expect the users of this method to recover from this exception by dynamically changing the node name and re-use the existing builder. Thus, it doesn't make much difference.

That said, I agree it is better to do the check before modifying the builders. 
I will make a change. 

Thanks

Reply:
see #1808